### ISSUE TITLE
display-name should warn about anonymous components inside higher-order functions

### ISSUE THREAD

**Issue by esamattis** [yannickcr/eslint-plugin-react#378]
Continuing from #367 but I'm opening a new issue because this is actually a different issue.

As I mentioned - in ES2015 `fn.name` is [picked up from the variable name](http://www.2ality.com/2015/09/function-names-es6.html) in statements like

``` js
const Foo = () => <div>foo</div>;
assert(Foo.name === "Foo");
```

and Babel handles [it](https://babeljs.io/repl/#?experimental=false&evaluate=false&loose=false&spec=false&code=const%20Foo%20%3D%20%28%29%20%3D%3E%20%3Cdiv%3Efoo%3C%2Fdiv%3E%3B).

But the variable name is not picked up when using higher-order functions:

``` js
const Foo = pure(() => <div>foo</div>);
```

See Babel output [here](https://babeljs.io/repl/#?experimental=false&evaluate=false&loose=false&spec=false&code=const%20Foo%20%3D%20pure%28%28%29%20%3D%3E%20%3Cdiv%3Efoo%3C%2Fdiv%3E%29%3B).

This is not warned by `react/display-name` and I think it should.

This is a pattern I've used with [recompose/pure](https://github.com/acdlite/recompose#optimize-rendering-performance) but I'd like avoid it future because it makes debugging harder.

So the preferred way to write this would be

``` js
let Foo = () => <div>foo</div>;
Foo = pure(Foo);
```

Or is it too hard to distinguish this from callback functions like

``` js
numbers.map(i => i*2);
```

?

**Comment by lencioni** [comment]
I'm not sure how feasible this is, since you likely will have iterators (e.g. map) returning JSX all over the place that you don't really need to assign to variables like this.

**Comment by ljharb** [comment]
Another alternative, if you want a reliable function name, would be to simply not use arrow functions - a normal named function will ensure you *always* have a name.

**Comment by travi** [comment]
until the function gets minified. then the name gets mangled

**Comment by ljharb** [comment]
@travi only if you've enabled function name mangling in uglify (or whatever tool you're using) - which you shouldn't do.

**Comment by travi** [comment]
thats fair, but since it's enabled by default in uglify (i think), its bitten me enough times to make sure the `displayName` is set explicitly.

**Comment by ljharb** [comment]
Increasing page weight and adding redundant code just to work around an unwise default in uglify is a really poor solution :-/

**Comment by travi** [comment]
yeah, i don't disagree. its more a matter of figuring out a fix when i didn't realize uglify allowed that configuration and not readdressing after learning about that. 

the page weight increase would be minimal (minified function name + displayName vs non-minified function name), but the redundant code and making sure that displayName doesnt get out of sync with the function name make it worth revisiting.

my intent of commenting wasnt really to suggest that this was the best approach, but more to clarify for someone else that might stumble on this thread that a named function might not _always_ have the name you expect, depending on how your build pipeline is configured. 

there are lots of tools to keep track of and keep their configuration in an optimal state. its very easy to be unaware of an option or not taken the time to correct it after learning about it, so hopefully this extra info helps someone get past mangled function names if they do happen to land here. i should have included the uglify config suggestion in my initial comment.

**Comment by ljharb** [comment]
You can achieve this best now with the `function-component-definition` rule, by prohibiting arrow function components entirely.
